dali/test/python/test_torch_pipeline_rnnt.py

```diff
@@ -355,77 +354,161 @@ def test_torch_vs_dali_normalize():
+        mel_spec = torch_mel_fbank(spec, npy_files_sr, device=device)
+        log_features = torch_log(mel_spec, device=device)
+        torch_out = torch_frame_splicing(log_features, stacking=stacking, subsampling=subsampling, device=device)
+        dali_out = dali_frame_splicing(log_features, stacking, subsampling, device=device)
+        np.testing.assert_allclose(torch_out, dali_out, atol=1e-4)
+
+def test_torch_vs_dali_frame_splicing():
+    for device in ['cpu', 'gpu']:
+        for stacking, subsampling in [(1, 1), (3, 1), (3, 2), (3, 3)]:
+            yield _testimpl_torch_vs_dali_frame_splicing, stacking, subsampling, device
+
 @pipeline_def(batch_size=1, device_id=0, num_threads=3)
-def rnnt_train_pipe(files, sample_rate=16000, silence_threshold=-80, preemph_coeff=.97,
+def rnnt_train_pipe(files, sample_rate, silence_threshold=-80, preemph_coeff=.97,
                     window_size=.02, window_stride=.01, window="hann", nfeatures=64, nfft=512, frame_splicing=1,
-                    lowfreq=0.0, highfreq=None, normalize_type='per_feature'):
+                    lowfreq=0.0, highfreq=None, normalize_type='per_feature', device='cpu'):
     norm_axes = [1] if 'per_feature' else [0, 1]
     win_len, win_hop = win_args(sample_rate, window_size, window_stride)
     window_fn = torch_windows.get(window, None)
     window_fn_arg = window_fn(win_len, periodic=False).detach().numpy().tolist() if window_fn else None
 
     data, _ = fn.readers.file(files=files, device="cpu", random_shuffle=False, shard_id=0, num_shards=1)
     audio, _ = fn.decoders.audio(data, dtype=types.FLOAT, downmix=True)
+
+    audio_shape = fn.shapes(audio, dtype=types.INT32)
+    audio_len = fn.slice(audio_shape, 0, 1, axes=(0,))
+    spec_len = audio_len // win_hop + 1
+
+    if device == 'gpu':
+        audio = audio.gpu()
+
     preemph_audio = fn.preemphasis_filter(audio, preemph_coeff=preemph_coeff, border='zero')
     spec = fn.spectrogram(preemph_audio, nfft=nfft, window_fn=window_fn_arg, window_length=win_len, window_step=win_hop,
                           center_windows=True, reflect_padding=True)
-
     mel_spec = fn.mel_filter_bank(spec, sample_rate=sample_rate, nfilter=nfeatures, freq_low=lowfreq, freq_high=highfreq)
     log_features = fn.to_decibels(mel_spec + 1e-20, multiplier=np.log(10), reference=1.0, cutoff_db=-80)
 
     if frame_splicing > 1:
-        log_features = fn.transpose(log_features, perm=[1, 0])
-        log_features = fn.reshape(log_features, rel_shape=[-1, frame_splicing])
-        log_features = fn.pad(log_features, axes=[0], fill_value=0, align=frame_splicing, shape=[1])
-        log_features = fn.transpose(log_features, perm=[1, 0])
+        seq = [log_features]
+        for n in range(1, frame_splicing):
+            f = fn.slice(log_features, start=n, axes=(1,))
+            f = fn.pad(f, shape=fn.cat(nfeatures, spec_len))
+            seq.append(f)
+        log_features_spliced = fn.cat(*seq, axis=0)
```

Comment:
I wonder how much of the GPU memory it would consume (as we add multiple slice and pad operators)?
When we have better memory management the buffer probably could be reused between instances, but not now.

Reply:
Now that I think of it, I don't need the pad. I can simply use slice with out-of-bounds policy set to pad.